### ISSUE TITLE
go/types: missing cycle errors

### ISSUE THREAD

**Issue by griesemer** [golang/go#34333]
```Go
package p

type A interface {
	B
}

type B interface {
	A
}
```
produces: `"illegal cycle in declaration of A"` with go/types. But when we introduce a method `m` such as in:
```Go
package p

type A interface {
	m() B
	B
}

type B interface {
	A
}
```
the error disappears and the code typechecks. If `m` is declared after the embedded interface `B`, the error appears again.

This is an ordering issue in `go/types`.

**Comment by griesemer** [comment]
PS: This bug was introduced with a80c5f0588e3835ff37c45a15d179571d9e403e0 .

**Comment by griesemer** [comment]
This is a general cycle detection bug. Replacing `interface` with `struct`:
```Go
package p

type A struct {
	f func() B
	b B
}

type B struct {
	A
}
```
type-checks but it shouldn't.

**Comment by av86743** [comment]
Root cause: 

Analysis of [detailed design proposal](https://github.com/golang/proposal/blob/master/design/6977-overlapping-interfaces.md) (Jul-Sep 2019) is not based on formal model.

**Comment by griesemer** [comment]
@av86743 Thanks, but the design proposal you're mentioning is not the root cause at all - it is in fact unrelated. It just so happens that I changed the implementation of interface type-checking entirely when implementing overlapping interfaces, which in turn exposed an underlying bug. As I have mentioned [here](https://github.com/golang/go/issues/34333#issuecomment-531995960), this bug also appears for structs which have nothing to do with interfaces; in fact this bug has been around for an embarrassingly long time.

**Comment by av86743** [comment]
@griesemer

> ... the design proposal you're mentioning is not the root cause at all ...

This is not what I said.  I said that "_Analysis of ... proposal is [essentially absent]_".  

> ... in fact this bug has been around for an embarrassingly long time.

I certainly appreciate your recognition of the fact.

**Comment by griesemer** [comment]
@av86743 Hm, apologies for the misunderstanding, but I don't understand what you are saying. Since the specific proposal is unrelated to this bug, I don't see why you bring in that proposal in the first place.

**Comment by av86743** [comment]
@griesemer

> ... I don't see why you bring in that proposal in the first place.

This [proposal](https://github.com/golang/proposal/blob/master/design/6977-overlapping-interfaces.md) is the first and only place where you specify (in section "Proposal") what you are intending to do. It also indicates a period when implementation has taken place.

**Comment by griesemer** [comment]
@av86743 Please see the updated title of this issue. It is unrelated to the [proposal](https://github.com/golang/proposal/blob/master/design/6977-overlapping-interfaces.md).

**Comment by av86743** [comment]
@griesemer 

Your [example with structs](https://play.golang.org/p/FKhsFNVMZ0R) does not type-check with current go1.12 (and likely with earlier go versions) - as it should.  It started to fail only after "unrelated" changes per said proposal.  Think a moment about it.

**Comment by griesemer** [comment]
@av86743 The latest released version is go1.13 (commit cc8838d645b2b7026c1f3aaceb011775c5ca3a08), and on my machine the struct example is (incorrectly) accepted. This code failed before said proposal was proposed or implemented.

**Comment by av86743** [comment]
@griesemer

> [edited]
> PS: This bug was ~introduced~ exposed with [a80c5f0](https://github.com/golang/go/commit/a80c5f0588e3835ff37c45a15d179571d9e403e0) .

Then what about this?

**Comment by griesemer** [comment]
@av86743 Initial guess/misdiagnosis - which is why I corrected it once I [realized](https://github.com/golang/go/issues/34333#issuecomment-531995960) that this was unrelated. The problem is due to how cycle detection was changed a while back. Anyway, I consider this discussed resolved at this point. Thanks.

**Comment by av86743** [comment]
"... bug was _exposed_ with [commit]" is rather strange choice of wording if you admit that regression in cycle detection appeared before commit was applied (or even existed.)  Clearly, the initial interface example is not part of the commit in question but was somehow arrived at, at a later time.  Which in turn translates into "...  bug was _not exposed_ with [commit]".  Hence the root cause.
